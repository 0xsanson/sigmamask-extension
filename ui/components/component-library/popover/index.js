export { Popover } from './popover';
export { POPOVER_POSITIONS } from './popover.constants';
