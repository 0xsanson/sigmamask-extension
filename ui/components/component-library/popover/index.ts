export { Popover } from './popover';
export { PopoverPosition } from './popover.constants';
export type { PopoverProps } from './popover.types';
