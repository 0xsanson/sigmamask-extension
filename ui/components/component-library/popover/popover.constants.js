export const POPOVER_POSITIONS = {
  // TODO: add positions
};
