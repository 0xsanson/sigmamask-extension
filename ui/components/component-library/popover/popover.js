import React, { useState } from 'react';
import { createPortal } from 'react-dom';
import { PropTypes } from 'prop-types';
import { usePopper } from 'react-popper';
import classnames from 'classnames';
import { BorderRadius, Color } from '../../../helpers/constants/design-system';
import Box from '../../ui/box/box';
import { Button } from '..';
import { PopoverPosition } from '.';

export const Popover = ({
  children,
  // content,
  position = PopoverPosition.bottom,
  hasArrow = true,
  onClick = false,
  onHover = true,
  onFocus = false,
  matchWidth = false,
  preventOverflow = false,
  flip = false,
  className,
  ...props
}) => {
  const [referenceElement, setReferenceElement] = useState(null);
  const [popperElement, setPopperElement] = useState(null);
  const [arrowElement, setArrowElement] = useState(null);
  const [isOpen, setIsOpen] = useState(false);
  // Define Popper options
  const { styles, attributes } = usePopper(referenceElement, popperElement, {
    placement: position,
    modifiers: [
      {
        name: 'preventOverflow',
        enabled: position === 'auto' ? true : preventOverflow,
      },
      {
        name: 'flip',
        enabled: position === 'auto' ? true : flip,
      },
      {
        name: 'arrow',
        enabled: hasArrow,
        options: {
          element: arrowElement,
        },
      },
      {
        name: 'offset',
        options: {
          offset: [0, 8],
        },
      },
    ],
  });
  // Define width to match reference element or auto
  const contentStyle = {
    width: matchWidth ? referenceElement?.clientWidth : 'auto',
  };

  const handleClick = () => {
    if (onClick) {
      setIsOpen(!isOpen);
    }
  };

  const handleMouseEnter = () => {
    if (onHover) {
      setIsOpen(true);
    }
  };

  const handleMouseLeave = () => {
    if (onHover) {
      setIsOpen(false);
    }
  };

  const handleFocus = () => {
    if (onFocus) {
      setIsOpen(true);
    }
  };

  const handleClose = () => {
    setIsOpen(false);
  };

  return (
    <>
      <div
        className="popover-reference"
        ref={setReferenceElement}
        onClick={handleClick}
        onMouseEnter={handleMouseEnter}
        onMouseLeave={handleMouseLeave}
        onFocus={handleFocus}
        onBlur={handleClose}
      >
        <Button>{children}</Button>
      </div>
      {isOpen &&
        createPortal(
          <Box
            borderColor={Color.borderDefault}
            borderRadius={BorderRadius.XL}
            backgroundColor={Color.backgroundDefault}
            padding={4}
            className={classnames('popover', { 'popover--open': isOpen })}
            ref={setPopperElement}
            style={{ ...styles.popper, ...contentStyle }}
            {...attributes.popper}
          >
            {children} - This is the popper content
            {hasArrow && (
              <Box
                borderColor={Color.borderDefault}
                className={classnames('arrow')}
                ref={setArrowElement}
                style={styles.arrow}
                {...attributes.arrow}
              />
            )}
          </Box>,
          document.body,
        )}
    </>
  );
};

Popover.propTypes = {
  /**
   * The children to be rendered inside the Popover
   */
  children: PropTypes.node,
  /**
   * An additional className to apply to the Popover.
   */
  className: PropTypes.string,
  /**
   * Popover accepts all the props from Box
   */
  ...Box.propTypes,
};
