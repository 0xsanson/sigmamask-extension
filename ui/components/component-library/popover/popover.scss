.mm-popover {
  // TODO
}

#arrow {
  display: inline-block;
  position: absolute;
  width: 8px;
  height: 8px;
  transform: rotate(-45deg);
  transform-origin: 100% 0;
}


#tooltip[data-popper-placement^='top'] > #arrow {
  bottom: -4px;
}

#tooltip[data-popper-placement^='bottom'] > #arrow {
  top: -4px;
}

#tooltip[data-popper-placement^='left'] > #arrow {
  right: -4px;
}

#tooltip[data-popper-placement^='right'] > #arrow {
  left: -4px;
}
