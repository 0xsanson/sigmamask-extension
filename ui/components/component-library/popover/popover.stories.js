import React, { useState, useEffect } from 'react';
import Box from '../../ui/box/box';
import {
  Color,
  AlignItems,
  DISPLAY,
  JustifyContent,
} from '../../../helpers/constants/design-system';
import { TextField } from '..';
import README from './README.mdx';
import { Popover, PopoverPosition } from '.';

const marginSizeControlOptions = [
  undefined,
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  'auto',
];

export default {
  title: 'Components/ComponentLibrary/Popover',
  component: Popover,
  parameters: {
    docs: {
      page: README,
    },
    controls: { sort: 'alpha' },
  },
  argTypes: {
    children: {
      control: 'text',
    },
    className: {
      control: 'text',
    },
    marginTop: {
      options: marginSizeControlOptions,
      control: 'select',
      table: { category: 'box props' },
    },
    marginRight: {
      options: marginSizeControlOptions,
      control: 'select',
      table: { category: 'box props' },
    },
    marginBottom: {
      options: marginSizeControlOptions,
      control: 'select',
      table: { category: 'box props' },
    },
    marginLeft: {
      options: marginSizeControlOptions,
      control: 'select',
      table: { category: 'box props' },
    },
  },
  args: {
    children: 'Popover',
  },
};

export const DefaultStory = (args) => {
  const [referenceElement, setReferenceElement] = useState();
  const [isOpen, setIsOpen] = useState(false);

  const handleClick = () => {
    setIsOpen(!isOpen);
  };

  const handleMouseEnter = () => {
    setIsOpen(true);
  };

  const handleMouseLeave = () => {
    setIsOpen(false);
  };

  const handleFocus = () => {
    setIsOpen(true);
  };

  const handleClose = () => {
    setIsOpen(false);
  };

  const handleKeyDown = (e) => {
    if (e.key === 'Escape') {
      setIsOpen(false);
    }
  };

  useEffect(() => {
    if (isOpen) {
      document.addEventListener('keydown', handleKeyDown);
    } else {
      document.removeEventListener('keydown', handleKeyDown);
    }
  }, [isOpen]);

  const setButtonRef = (ref) => {
    setReferenceElement(ref);
  };

  return (
    <Box
      style={{ height: '200vh', width: '200vw' }}
      display={DISPLAY.FLEX}
      justifyContent={JustifyContent.center}
      alignItems={AlignItems.center}
    >
      <Box
        ref={setButtonRef}
        // onClick={handleClick}
        // onMouseEnter={handleMouseEnter}
        // onMouseLeave={handleMouseLeave}
        onFocus={handleFocus}
        onBlur={handleClose}
        // backgroundColor={Color.primaryDefault}
        // style={{ width: 200, height: 200 }}
        as="input"
        type="text"
        placeholder="write a description here about a popover"
      ></Box>
      <Popover
        position={PopoverPosition.bottomStart}
        referenceElement={referenceElement}
        isOpen={isOpen}
        // hasArrow
        matchWidth
        onClose={handleClose}
        onBack={() => console.log('back')}
        {...args}
      >
        write a description here about a popover
      </Popover>
    </Box>
  );
};

DefaultStory.storyName = 'Default';
