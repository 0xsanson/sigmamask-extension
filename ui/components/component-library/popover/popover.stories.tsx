import React, { useState, useEffect } from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import Box from '../../ui/box/box';
import {
  AlignItems,
  BackgroundColor,
  BorderColor,
  Color,
  DISPLAY,
  JustifyContent,
  TextVariant,
  TEXT_ALIGN,
} from '../../../helpers/constants/design-system';
import { Icon, IconName, IconSize, PopoverHeader, Text } from '..';
import README from './README.mdx';
import { Popover, PopoverPosition } from '.';

export default {
  title: 'Components/ComponentLibrary/Popover',
  component: Popover,
  parameters: {
    docs: {
      page: README,
    },
  },
  argTypes: {
    children: {
      control: 'text',
    },

    position: {
      options: PopoverPosition,
      control: 'select',
    },
    className: {
      control: 'text',
    },
  },
  args: {
    children: 'Popover',
  },
} as ComponentMeta<typeof Popover>;

const Template: ComponentStory<typeof Popover> = (args) => {
  const [referenceElement, setReferenceElement] = useState();
  const [isOpen, setIsOpen] = useState(true);

  const handleClick = () => {
    setIsOpen(!isOpen);
  };

  // Example of how to use mouse events to open and close popover
  // const handleMouseEnter = () => {
  //   setIsOpen(true);
  // };

  // const handleMouseLeave = () => {
  //   setIsOpen(false);
  // };

  // Example of how open popover with focus
  // pair with onBlur (example using handleClose) to close popover
  // const handleFocus = () => {
  //   setIsOpen(true);
  // };

  const handleClose = () => {
    setIsOpen(false);
  };

  const handleKeyDown = (e) => {
    if (e.key === 'Escape') {
      setIsOpen(false);
    }
  };

  // Example of how to use keyboard events to close popover with escape key
  useEffect(() => {
    if (isOpen) {
      document.addEventListener('keydown', handleKeyDown);
    } else {
      document.removeEventListener('keydown', handleKeyDown);
    }
  }, [isOpen]);

  // Example of how to use ref to open popover
  const setButtonRef = (ref) => {
    setReferenceElement(ref);
  };

  return (
    <Box
      style={{ height: '200vh', width: '200vw' }}
      display={DISPLAY.FLEX}
      justifyContent={JustifyContent.center}
      alignItems={AlignItems.center}
    >
      <Box
        ref={setButtonRef}
        onClick={handleClick}
        // onMouseEnter={handleMouseEnter}
        // onMouseLeave={handleMouseLeave}
        // onFocus={handleFocus}
        // onBlur={handleClose}
        backgroundColor={BackgroundColor.primaryAlternative}
        style={{ width: 200, height: 200 }}
        color={Color.primaryInverse}
        as="button"
      >
        Click to toggle popover
      </Box>
      {isOpen && (
        <Popover
          position={PopoverPosition.BottomStart}
          referenceElement={referenceElement}
          isOpen={isOpen}
          hasArrow
          // matchWidth
          {...args}
        >
          write a description here about a popover
        </Popover>
      )}
    </Box>
  );
};

export const DefaultStory = Template.bind({});
DefaultStory.storyName = 'Default';

export const ReferenceElement = ({ args }) => {
  const [referenceElement, setReferenceElement] = useState();

  const setButtonRef = (ref) => {
    setReferenceElement(ref);
  };

  return (
    <>
      <Box
        ref={setButtonRef}
        backgroundColor={BackgroundColor.primaryDefault}
        style={{ width: 200, height: 200 }}
      ></Box>
      <Popover
        position={PopoverPosition.Bottom}
        referenceElement={referenceElement}
        isOpen={true}
        hasArrow
        {...args}
      >
        <Text>Reference Element</Text>
      </Popover>
    </>
  );
};

export const Children = ({ args }) => {
  const [referenceElement, setReferenceElement] = useState();

  const setButtonRef = (ref) => {
    setReferenceElement(ref);
  };

  return (
    <>
      <Box
        ref={setButtonRef}
        backgroundColor={BackgroundColor.primaryDefault}
        style={{ width: 200, height: 200 }}
      ></Box>
      <Popover
        referenceElement={referenceElement}
        isOpen={true}
        hasArrow
        {...args}
      >
        <Text>
          Demo of popover with children.{' '}
          <Icon size={IconSize.Inherit} name={IconName.Info} />
        </Text>
        <Text variant={TextVariant.bodySm}>Control the content</Text>
      </Popover>
    </>
  );
};

export const Position = ({ args }) => {
  const [referenceElement, setReferenceElement] = useState();
  const [referenceAutoElement, setReferenceAutoElement] = useState();

  const setButtonRef = (ref) => {
    setReferenceElement(ref);
  };

  const setRefAuto = (ref) => {
    setReferenceAutoElement(ref);
  };

  return (
    <>
      <Box
        style={{
          width: '90vw',
          minWidth: '650px',
          height: '90vh',
          minHeight: '400px',
        }}
        borderColor={BorderColor.borderDefault}
        display={DISPLAY.FLEX}
        justifyContent={JustifyContent.center}
        alignItems={AlignItems.center}
        marginBottom={4}
      >
        <Box
          ref={setButtonRef}
          backgroundColor={BackgroundColor.primaryMuted}
          style={{ width: 400, height: 200 }}
          display={DISPLAY.FLEX}
          justifyContent={JustifyContent.center}
          alignItems={AlignItems.center}
          textAlign={TEXT_ALIGN.CENTER}
        >
          Position
        </Box>
        <Popover
          position={PopoverPosition.TopStart}
          referenceElement={referenceElement}
          isOpen={true}
          hasArrow
          {...args}
        >
          {PopoverPosition.TopStart}
        </Popover>
        <Popover
          position={PopoverPosition.Top}
          referenceElement={referenceElement}
          isOpen={true}
          hasArrow
          {...args}
        >
          {PopoverPosition.Top}
        </Popover>
        <Popover
          position={PopoverPosition.TopEnd}
          referenceElement={referenceElement}
          isOpen={true}
          hasArrow
          {...args}
        >
          {PopoverPosition.TopEnd}
        </Popover>
        <Popover
          position={PopoverPosition.RightStart}
          referenceElement={referenceElement}
          isOpen={true}
          hasArrow
          {...args}
        >
          {PopoverPosition.RightStart}
        </Popover>
        <Popover
          position={PopoverPosition.Right}
          referenceElement={referenceElement}
          isOpen={true}
          hasArrow
          {...args}
        >
          {PopoverPosition.Right}
        </Popover>
        <Popover
          position={PopoverPosition.RightEnd}
          referenceElement={referenceElement}
          isOpen={true}
          hasArrow
          {...args}
        >
          {PopoverPosition.RightEnd}
        </Popover>
        <Popover
          position={PopoverPosition.BottomStart}
          referenceElement={referenceElement}
          isOpen={true}
          hasArrow
          {...args}
        >
          {PopoverPosition.BottomStart}
        </Popover>
        <Popover
          position={PopoverPosition.Bottom}
          referenceElement={referenceElement}
          isOpen={true}
          hasArrow
          {...args}
        >
          {PopoverPosition.Bottom}
        </Popover>
        <Popover
          position={PopoverPosition.BottomEnd}
          referenceElement={referenceElement}
          isOpen={true}
          hasArrow
          {...args}
        >
          {PopoverPosition.BottomEnd}
        </Popover>
        <Popover
          position={PopoverPosition.LeftStart}
          referenceElement={referenceElement}
          isOpen={true}
          hasArrow
          {...args}
        >
          {PopoverPosition.LeftStart}
        </Popover>
        <Popover
          position={PopoverPosition.Left}
          referenceElement={referenceElement}
          isOpen={true}
          hasArrow
          {...args}
        >
          {PopoverPosition.Left}
        </Popover>
        <Popover
          position={PopoverPosition.LeftEnd}
          referenceElement={referenceElement}
          isOpen={true}
          hasArrow
          {...args}
        >
          {PopoverPosition.LeftEnd}
        </Popover>
      </Box>
      <Box
        style={{
          width: '90vw',
          minWidth: '650px',
          height: '90vh',
          minHeight: '400px',
          overflow: 'scroll',
        }}
        borderColor={BorderColor.borderDefault}
      >
        <Box
          style={{
            width: '200vw',
            height: '200vh',
          }}
          display={DISPLAY.FLEX}
          justifyContent={JustifyContent.center}
          alignItems={AlignItems.center}
        >
          <Box
            ref={setRefAuto}
            backgroundColor={BackgroundColor.primaryMuted}
            style={{ width: 400, height: 200 }}
            display={DISPLAY.FLEX}
            justifyContent={JustifyContent.center}
            alignItems={AlignItems.center}
            textAlign={TEXT_ALIGN.CENTER}
          >
            Position
          </Box>
          <Popover
            position={PopoverPosition.Auto}
            referenceElement={referenceAutoElement}
            isOpen={true}
            hasArrow
            {...args}
          >
            {PopoverPosition.Auto}
          </Popover>
        </Box>
      </Box>
    </>
  );
};

export const HasArrow = ({ args }) => {
  const [referenceElement, setReferenceElement] = useState();

  const setButtonRef = (ref) => {
    setReferenceElement(ref);
  };

  return (
    <>
      <Box
        ref={setButtonRef}
        backgroundColor={BackgroundColor.primaryDefault}
        style={{ width: 200, height: 200 }}
      ></Box>
      <Popover
        position={PopoverPosition.RightStart}
        referenceElement={referenceElement}
        isOpen={true}
        hasArrow
        {...args}
      >
        <Text>Popover with arrow</Text>
      </Popover>
      <Popover
        position={PopoverPosition.RightEnd}
        referenceElement={referenceElement}
        isOpen={true}
        {...args}
      >
        <Text>Popover with no arrow</Text>
      </Popover>
    </>
  );
};

export const IsOpen = ({ args }) => {
  const [referenceElement, setReferenceElement] = useState();
  const [isOpen, setIsOpen] = useState(true);

  const setButtonRef = (ref) => {
    setReferenceElement(ref);
  };

  const handleClick = () => {
    setIsOpen(!isOpen);
  };

  return (
    <>
      <Box
        ref={setButtonRef}
        backgroundColor={BackgroundColor.primaryMuted}
        style={{ width: 200, height: 200 }}
        onClick={handleClick}
        display={DISPLAY.FLEX}
        justifyContent={JustifyContent.center}
        alignItems={AlignItems.center}
      >
        Click to toggle popover
      </Box>

      <Popover
        position={PopoverPosition.RightStart}
        referenceElement={referenceElement}
        isOpen={true}
        hasArrow
        {...args}
      >
        <Text>isOpen always true</Text>
      </Popover>
      {isOpen && (
        <Popover
          position={PopoverPosition.RightEnd}
          referenceElement={referenceElement}
          hasArrow
          isOpen={isOpen}
          {...args}
        >
          <Text>isOpen tied to boolean</Text>
        </Popover>
      )}
    </>
  );
};

export const Flip = ({ args }) => {
  const [referenceElement, setReferenceElement] = useState();

  const setButtonRef = (ref) => {
    setReferenceElement(ref);
  };

  return (
    <Box
      style={{ height: '200vh' }}
      display={DISPLAY.FLEX}
      justifyContent={JustifyContent.center}
      alignItems={AlignItems.center}
    >
      <Box
        ref={setButtonRef}
        backgroundColor={BackgroundColor.primaryMuted}
        style={{ width: 200, height: 200 }}
        display={DISPLAY.FLEX}
        justifyContent={JustifyContent.center}
        alignItems={AlignItems.center}
      >
        Scroll to see popover flip
      </Box>
      <Popover
        position={PopoverPosition.TopStart}
        referenceElement={referenceElement}
        isOpen={true}
        hasArrow
        {...args}
      >
        false
      </Popover>
      <Popover
        position={PopoverPosition.TopEnd}
        referenceElement={referenceElement}
        hasArrow
        flip
        isOpen={true}
        {...args}
      >
        true
      </Popover>
    </Box>
  );
};

export const PreventOverflow = ({ args }) => {
  const [referenceElement, setReferenceElement] = useState();

  const setButtonRef = (ref) => {
    setReferenceElement(ref);
  };

  return (
    <Box
      style={{ height: '200vh', width: '100vw' }}
      display={DISPLAY.FLEX}
      justifyContent={JustifyContent.center}
      alignItems={AlignItems.center}
    >
      <Box
        ref={setButtonRef}
        backgroundColor={BackgroundColor.primaryMuted}
        style={{ width: 200, height: 200 }}
        display={DISPLAY.FLEX}
        justifyContent={JustifyContent.center}
        alignItems={AlignItems.center}
        textAlign={TEXT_ALIGN.CENTER}
      >
        Scroll to see popover preventOverflow
      </Box>
      <Popover
        position={PopoverPosition.Left}
        referenceElement={referenceElement}
        isOpen={true}
        hasArrow
        {...args}
      >
        false
      </Popover>
      <Popover
        position={PopoverPosition.Right}
        referenceElement={referenceElement}
        hasArrow
        preventOverflow
        isOpen={true}
        {...args}
      >
        true
      </Popover>
    </Box>
  );
};

export const ReferenceHidden = ({ args }) => {
  const [referenceElement, setReferenceElement] = useState();

  const setButtonRef = (ref) => {
    setReferenceElement(ref);
  };

  return (
    <Box
      style={{ height: '200vh', width: '100vw' }}
      display={DISPLAY.FLEX}
      justifyContent={JustifyContent.center}
    >
      <Box
        ref={setButtonRef}
        backgroundColor={BackgroundColor.primaryMuted}
        style={{ width: 200, height: 200 }}
        display={DISPLAY.FLEX}
        justifyContent={JustifyContent.center}
        alignItems={AlignItems.center}
        textAlign={TEXT_ALIGN.CENTER}
      >
        Scroll to see popover referenceHidden
      </Box>
      <Popover
        position={PopoverPosition.BottomStart}
        referenceElement={referenceElement}
        isOpen={true}
        hasArrow
        {...args}
      >
        <Text>false</Text>
      </Popover>
      <Popover
        position={PopoverPosition.BottomEnd}
        referenceElement={referenceElement}
        hasArrow
        referenceHidden
        isOpen={true}
        {...args}
      >
        <Text>true</Text>
      </Popover>
    </Box>
  );
};

export const MatchWidth = ({ args }) => {
  const [referenceElement, setReferenceElement] = useState();

  const setButtonRef = (ref) => {
    setReferenceElement(ref);
  };

  return (
    <>
      <Box
        ref={setButtonRef}
        backgroundColor={BackgroundColor.primaryDefault}
        style={{ width: 200, height: 200 }}
      ></Box>
      <Popover
        position={PopoverPosition.Bottom}
        referenceElement={referenceElement}
        isOpen={true}
        matchWidth
        {...args}
      >
        <Text>
          Setting matchWidth to true will make the popover match the width of
          the reference element
        </Text>
      </Popover>
    </>
  );
};

export const WithPopoverHeader = ({ args }) => {
  const [refTitleElement, setRefTitleElement] = useState();

  const setButtonRef = (ref) => {
    setRefTitleElement(ref);
  };

  return (
    <>
      <Box
        ref={setButtonRef}
        backgroundColor={BackgroundColor.primaryDefault}
        style={{ width: 200, height: 200 }}
      ></Box>
      <Popover
        referenceElement={refTitleElement}
        // backgroundColor={BackgroundColor.primaryDefault}
        // color={Color.primaryInverse}
        // borderColor={BorderColor.primaryDefault}
        // arrowProps={{ borderColor: BorderColor.primaryDefault }}
        isOpen={true}
        hasArrow
        {...args}
      >
        <PopoverHeader
          onClose={() => console.log('close')}
          onBack={() => console.log('back')}
          color={Color.inherit}
        >
          Popover Title
        </PopoverHeader>
        Title should be short and concise. It should be sentence case and no
        period.
      </Popover>
    </>
  );
};
