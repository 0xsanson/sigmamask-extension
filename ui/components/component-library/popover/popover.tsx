import React, { useState, forwardRef, useImperativeHandle, Ref } from 'react';
import { createPortal } from 'react-dom';
import { usePopper, Placement } from 'react-popper';
import classnames from 'classnames';
import {
  AlignItems,
  BorderRadius,
  Color,
  DISPLAY,
  JustifyContent,
  Size,
  TextVariant,
  TEXT_ALIGN,
} from '../../../helpers/constants/design-system';
import Box from '../../ui/box/box';
import { ButtonIcon, ICON_NAMES, Text } from '..';
import { PopoverPosition, PopoverProps } from '.';

type PopoverPosition = Placement;

interface PopoverRef {
  closePopover: () => void;
}

export const Popover = forwardRef<PopoverRef, PopoverProps>(
  (
    {
      children,
      className,
      position = PopoverPosition.Auto,
      hasArrow = false,
      matchWidth,
      preventOverflow = false,
      flip = false,
      referenceElement,
      isOpen,
      title,
      onClose,
      closeButtonProps,
      onBack,
      backButtonProps,
      ...props
    }: PopoverProps,
    ref: Ref<PopoverRef>,
  ) => {
    const [popperElement, setPopperElement] = useState<HTMLElement | null>(
      null,
    );
    const [arrowElement, setArrowElement] = useState<HTMLElement | null>(null);

    // Define Popper options
    const { styles, attributes } = usePopper(referenceElement, popperElement, {
      placement: position,
      modifiers: [
        {
          name: 'preventOverflow',
          enabled: position === PopoverPosition.Auto ? true : preventOverflow,
        },
        {
          name: 'flip',
          enabled: position === PopoverPosition.Auto ? true : flip,
        },
        {
          name: 'arrow',
          enabled: hasArrow,
          options: {
            element: arrowElement,
          },
        },
        {
          name: 'offset',
          options: {
            offset: [0, 8],
          },
        },
      ],
    });

    // Define width to match reference element or auto
    const contentStyle = {
      width: matchWidth ? referenceElement?.clientWidth : 'auto',
    };

    // Forwarding function to close the popover using the ref
    useImperativeHandle(ref, () => ({
      closePopover: () => {
        // Your close popover implementation here
        console.log('closePopover called');
      },
    }));

    return (
      <>
        {isOpen &&
          createPortal(
            <Box
              borderColor={Color.borderDefault}
              borderRadius={BorderRadius.XL}
              backgroundColor={Color.backgroundDefault}
              padding={4}
              className={classnames(
                'mm-popover',
                { 'mm-popover--open': isOpen },
                className,
              )}
              ref={setPopperElement}
              style={{ ...styles.popper, ...contentStyle }}
              {...attributes.popper}
              {...props}
            >
              {/* TODO: Replace with HeaderBase  Start */}
              <Box
                className="mm-popover__header"
                paddingLeft={onBack || onClose ? 8 : 0}
                paddingRight={onBack || onClose ? 8 : 0}
              >
                {onBack && (
                  <ButtonIcon
                    iconName={ICON_NAMES.ARROW_LEFT}
                    className="mm-popover__header-button-back"
                    size={Size.SM}
                    ariaLabel="back"
                    onClick={onBack}
                    {...backButtonProps}
                  />
                )}

                <Text
                  variant={TextVariant.headingSm}
                  className="mm-popover__header-title"
                  textAlign={TEXT_ALIGN.CENTER}
                >
                  {title}
                </Text>

                {onClose && (
                  <ButtonIcon
                    iconName={ICON_NAMES.CLOSE}
                    className="mm-popover__header-button-close"
                    size={Size.SM}
                    ariaLabel="close"
                    onClick={onClose}
                    {...closeButtonProps}
                  />
                )}
              </Box>
              {/* TODO: Replace with HeaderBase END */}
              {children}
              {hasArrow && (
                <Box
                  borderColor={Color.borderDefault}
                  className={classnames('mm-popover__arrow')}
                  ref={setArrowElement}
                  display={DISPLAY.FLEX}
                  justifyContent={JustifyContent.center}
                  alignItems={AlignItems.center}
                  style={styles.arrow}
                  {...attributes.arrow}
                />
              )}
            </Box>,
            document.body,
          )}
      </>
    );
  },
);
