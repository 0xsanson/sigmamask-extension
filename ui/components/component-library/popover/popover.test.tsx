/* eslint-disable jest/require-top-level-describe */
import { render } from '@testing-library/react';
import React from 'react';
import { Popover } from './popover';
import { PopoverPosition } from './popover.constants';

describe('Popover', () => {
  it('should render popover element correctly', () => {
    const { getByTestId, getByText, container } = render(
      <Popover data-testid="popover" isOpen={true}>
        Popover
      </Popover>,
    );
    expect(getByText('Popover')).toBeDefined();
    expect(container.querySelector('popover')).toBeDefined();
    expect(getByTestId('popover')).toHaveClass('mm-popover');
    expect(container).toMatchSnapshot();
  });

  it('should render popover title', () => {
    const { getByText } = render(
      <Popover data-testid="popover" title="Popover test" isOpen={true}>
        Popover
      </Popover>,
    );
    expect(getByText('Popover test')).toBeDefined();
    expect(getByText('Popover test')).toHaveClass('mm-popover__header-title');
  });

  it('should render popover position', () => {
    const { getByText } = render(
      <>
        <Popover
          isOpen={true}
          data-testid={PopoverPosition.Auto}
          position={PopoverPosition.Auto}
        >
          {PopoverPosition.Auto}
        </Popover>

        <Popover
          isOpen={true}
          data-testid={PopoverPosition.AutoStart}
          position={PopoverPosition.AutoStart}
        >
          {PopoverPosition.AutoStart}
        </Popover>

        <Popover
          isOpen={true}
          data-testid={PopoverPosition.AutoEnd}
          position={PopoverPosition.AutoEnd}
        >
          {PopoverPosition.AutoEnd}
        </Popover>

        <Popover
          isOpen={true}
          data-testid={PopoverPosition.Top}
          position={PopoverPosition.Top}
        >
          {PopoverPosition.Top}
        </Popover>

        <Popover
          isOpen={true}
          data-testid={PopoverPosition.TopStart}
          position={PopoverPosition.TopStart}
        >
          {PopoverPosition.TopStart}
        </Popover>

        <Popover
          isOpen={true}
          data-testid={PopoverPosition.TopEnd}
          position={PopoverPosition.TopEnd}
        >
          {PopoverPosition.TopEnd}
        </Popover>

        <Popover
          isOpen={true}
          data-testid={PopoverPosition.Right}
          position={PopoverPosition.Right}
        >
          {PopoverPosition.Right}
        </Popover>

        <Popover
          isOpen={true}
          data-testid={PopoverPosition.RightStart}
          position={PopoverPosition.RightStart}
        >
          {PopoverPosition.RightStart}
        </Popover>

        <Popover
          isOpen={true}
          data-testid={PopoverPosition.RightEnd}
          position={PopoverPosition.RightEnd}
        >
          {PopoverPosition.RightEnd}
        </Popover>

        <Popover
          isOpen={true}
          data-testid={PopoverPosition.Bottom}
          position={PopoverPosition.Bottom}
        >
          {PopoverPosition.Bottom}
        </Popover>

        <Popover
          isOpen={true}
          data-testid={PopoverPosition.BottomStart}
          position={PopoverPosition.BottomStart}
        >
          {PopoverPosition.BottomStart}
        </Popover>

        <Popover
          isOpen={true}
          data-testid={PopoverPosition.BottomEnd}
          position={PopoverPosition.BottomEnd}
        >
          {PopoverPosition.BottomEnd}
        </Popover>

        <Popover
          isOpen={true}
          data-testid={PopoverPosition.Left}
          position={PopoverPosition.Left}
        >
          {PopoverPosition.Left}
        </Popover>

        <Popover
          isOpen={true}
          data-testid={PopoverPosition.LeftStart}
          position={PopoverPosition.LeftStart}
        >
          {PopoverPosition.LeftStart}
        </Popover>

        <Popover
          isOpen={true}
          data-testid={PopoverPosition.LeftEnd}
          position={PopoverPosition.LeftEnd}
        >
          {PopoverPosition.LeftEnd}
        </Popover>
      </>,
    );

    expect(getByText(PopoverPosition.Auto)).toBeDefined();
    expect(getByText(PopoverPosition.AutoStart)).toBeDefined();
    expect(getByText(PopoverPosition.AutoEnd)).toBeDefined();
    expect(getByText(PopoverPosition.Top)).toBeDefined();
    expect(getByText(PopoverPosition.TopStart)).toBeDefined();
    expect(getByText(PopoverPosition.TopEnd)).toBeDefined();
    expect(getByText(PopoverPosition.Right)).toBeDefined();
    expect(getByText(PopoverPosition.RightStart)).toBeDefined();
    expect(getByText(PopoverPosition.RightEnd)).toBeDefined();
    expect(getByText(PopoverPosition.Bottom)).toBeDefined();
    expect(getByText(PopoverPosition.BottomStart)).toBeDefined();
    expect(getByText(PopoverPosition.BottomEnd)).toBeDefined();
    expect(getByText(PopoverPosition.Left)).toBeDefined();
    expect(getByText(PopoverPosition.LeftStart)).toBeDefined();
    expect(getByText(PopoverPosition.LeftEnd)).toBeDefined();
  });
});
