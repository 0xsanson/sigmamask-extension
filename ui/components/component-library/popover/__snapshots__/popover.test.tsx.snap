// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover should render popover element correctly 1`] = `<div />`;
